static/js/components/dashboard/CourseListCard.js

```diff
@@ -11,54 +11,145 @@ import type {
 } from '../../flow/couponTypes';
 import CourseRow from './CourseRow';
 import FinancialAidCalculator from '../../containers/FinancialAidCalculator';
+import type { CoursePrice } from '../../flow/dashboardTypes';
+import type { CourseRun } from '../../flow/programTypes';
+import {
+  FA_TERMINAL_STATUSES,
+  COUPON_CONTENT_TYPE_PROGRAM,
+  COUPON_AMOUNT_TYPE_PERCENT_DISCOUNT,
+  COUPON_AMOUNT_TYPE_FIXED_DISCOUNT,
+} from '../../constants';
+import {
+  makeAmountMessage,
+  makeCouponReason,
+} from '../../lib/coupon';
+import { pickExistingProps } from '../../util/util';
 
 export default class CourseListCard extends React.Component {
   props: {
-    coupon:                       ?Coupon,
+    coupon?:                      Coupon,
     program:                      Program,
-    courseEnrollAddStatus?:       string,
+    coursePrice:                  CoursePrice,
     prices:                       CalculatedPrices,
     openFinancialAidCalculator?:  () => void,
     now?:                         Object,
-    addCourseEnrollment:          (courseId: string) => void,
-    openCourseContactDialog:      (course: Course, canContactCourseTeam: boolean) => void
+    addCourseEnrollment:          (courseId: string) => Promise<*>,
+    openCourseContactDialog:      (course: Course, canContactCourseTeam: boolean) => void,
+    setEnrollSelectedCourseRun:   (r: CourseRun) => void,
+    setEnrollCourseDialogVisibility: (bool: boolean) => void,
   };
 
+  renderPriceMessage() {
+    const {
+      program,
+      coupon,
+      prices,
+    } = this.props;
+
+    // Special case: 100% off coupon
+    if (coupon &&
+        coupon.content_type === COUPON_CONTENT_TYPE_PROGRAM &&
+        coupon.amount_type === COUPON_AMOUNT_TYPE_PERCENT_DISCOUNT &&
+        coupon.amount.equals(1)
```

Comment:
Might be worth factoring out `isFree` from here: https://github.com/mitodl/micromasters/pull/2519/files#diff-e7226b0752d5be35f40e935a28554779R81 since we have this in two places now (albeit this additionally checks if it's a program coupon).